nexus/src/app/external_endpoints.rs

```diff
@@ -0,0 +1,1052 @@
+        // that doesn't appear to be compatible with the SNI value (DNS domain)
+        // that this request came in on.
+        //
+        // IMPORTANT: If we ever decide to do those things, they should only be
+        // used to decide which of several certificates is preferred.  We should
+        // always pick a certificate if we possibly can, even if it seems to be
+        // invalid.  A client can always choose not to trust it.  But in the
+        // unfortunate case where there are no good certificates, a customer's
+        // only option may be to instruct their client to trust an invalid
+        // certificate _so that they can log in and fix the certificate
+        // problem_.  If we provide no certificate at all here, a customer may
+        // have no way to fix the problem.
+        //
+        // Anyway, we don't yet do anything of these things.  For now, pick the
+        // certificate chain whose leaf certificate has the latest expiration
+        // time.
+
+        // This would be cleaner if Asn1Time impl'd Ord or even just a way to
+        // convert it to a Unix timestamp or any other comparable timestamp.
+        let mut latest_expiration: Option<&TlsCertificate> = None;
+        for t in &self.tls_certs {
+            // We'll choose this certificate (so far) if we find that it's
+            // anything other than "earlier" than the best we've seen so far.
+            // That includes the case where we haven't seen any so far, where
+            // this one is greater than or equal to the best so far, as well as
+            // the case where they're incomparable for whatever reason.  (This
+            // ensures that we always pick at least one.)
+            if latest_expiration.is_none()
+                || !matches!(
+                    t.parsed.not_after().partial_cmp(
+                        latest_expiration.unwrap().parsed.not_after()
+                    ),
+                    Some(std::cmp::Ordering::Less)
+                )
+            {
+                latest_expiration = Some(t);
+            }
+        }
+
+        latest_expiration.ok_or_else(|| {
+            anyhow!("silo {} has no usable certificates", self.silo_id)
+        })
+    }
+}
+
+/// Describes a problem encountered while assembling an [`ExternalEndpoints`]
+/// object
+#[derive(Clone, Debug, Error, SerializeDisplay)]
+enum ExternalEndpointError {
+    #[error(
+        "ignoring silo {dup_silo_id} ({dup_silo_name:?}): has the same DNS \
+        name ({dns_name:?}) as previously-found silo {first_silo_id} \
+        ({first_silo_name:?})"
+    )]
+    DupDnsName {
+        dup_silo_id: Uuid,
+        dup_silo_name: String,
+        first_silo_id: Uuid,
+        first_silo_name: String,
+        dns_name: String,
+    },
+
+    #[error("ignoring certificate for silo {silo_id}: {reason:#}")]
+    BadCert {
+        silo_id: Uuid,
+        #[source]
+        reason: Arc<anyhow::Error>,
```

Comment:
The short answer is that I wanted this structure to impl `Clone` and `anyhow::Error` doesn't impl `Clone`.  The reason I wanted this structure to impl `Clone` is so that `ExternalEndpoints` can impl `Clone`.  That's useful because it gets sent over a watch channel.  Borrows from a watch channel block writers.  So it's useful to be able to clone the thing you receive from it and then drop the borrow.